rpc/src/client/subscription.rs

```diff
@@ -0,0 +1,591 @@
+            .send(TerminateSubscription {
+                id: self.id.clone(),
+                query: self.query.clone(),
+                result_tx,
+            })
+            .await?;
+        result_rx.recv().await.ok_or_else(|| {
+            Error::client_internal_error(
+                "failed to hear back from subscription termination request".to_string(),
+            )
+        })?
+    }
+}
+
+/// A message sent to the subscription driver to terminate the subscription
+/// with the given parameters.
+///
+/// We expect the driver to use the `result_tx` channel to communicate the
+/// result of the termination request to the original caller.
+#[derive(Debug, Clone)]
+pub struct TerminateSubscription {
+    pub id: SubscriptionId,
+    pub query: String,
+    pub result_tx: ChannelTx<Result<()>>,
+}
+
+/// Each new subscription is automatically assigned an ID.
+///
+/// By default, we generate random [UUIDv4] IDs for each subscription to
+/// minimize chances of collision.
+///
+/// [UUIDv4]: https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct SubscriptionId(String);
+
+impl Default for SubscriptionId {
+    fn default() -> Self {
+        let mut bytes = [0; 16];
+        getrandom(&mut bytes).expect("RNG failure!");
+
+        let uuid = uuid::Builder::from_bytes(bytes)
+            .set_variant(uuid::Variant::RFC4122)
+            .set_version(uuid::Version::Random)
+            .build();
+
+        Self(uuid.to_string())
+    }
+}
+
+impl std::fmt::Display for SubscriptionId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl Into<Id> for SubscriptionId {
+    fn into(self) -> Id {
+        Id::Str(self.0)
+    }
+}
+
+impl TryInto<SubscriptionId> for Id {
+    type Error = Error;
+
+    fn try_into(self) -> std::result::Result<SubscriptionId, Self::Error> {
+        match self {
+            Id::Str(s) => Ok(SubscriptionId(s)),
+            Id::Num(i) => Ok(SubscriptionId(format!("{}", i))),
+            Id::None => Err(Error::client_internal_error(
+                "cannot convert an empty JSON-RPC ID into a subscription ID",
+            )),
+        }
+    }
+}
+
+impl AsRef<str> for SubscriptionId {
+    fn as_ref(&self) -> &str {
+        self.0.as_ref()
+    }
+}
+
+impl From<&str> for SubscriptionId {
+    fn from(s: &str) -> Self {
+        Self(s.to_string())
+    }
+}
+
+#[derive(Debug)]
+struct PendingSubscribe {
+    id: SubscriptionId,
+    query: String,
+    event_tx: ChannelTx<Result<Event>>,
+    result_tx: ChannelTx<Result<()>>,
+}
+
+#[derive(Debug)]
+struct PendingUnsubscribe {
+    id: SubscriptionId,
+    query: String,
+    result_tx: ChannelTx<Result<()>>,
+}
+
+/// The current state of a subscription.
+#[derive(Debug, Clone, PartialEq)]
+pub enum SubscriptionState {
+    Pending,
+    Active,
+    Cancelling,
+    NotFound,
```

Comment:
what does not found mean in this context? how's subscription state can be not found?

Reply:
Good point! It probably makes more sense to make this method return an optional subscription state: https://github.com/informalsystems/tendermint-rs/pull/516/files#diff-89fca07be889575fefb9e77a8de9051dR394